ethstorage/storage_manager.go

```diff
@@ -293,6 +322,175 @@ func (s *StorageManager) syncCheck(kvIdx uint64) error {
+		rangeStart := taskIdx * rangeSize
+		rangeEnd := (taskIdx + 1) * rangeSize
+		if taskIdx == taskNum - 1 {
+			rangeEnd = to
+		}
+		wg.Add(1)
+
+		go func(start, end, taskId uint64, out chan<- error) {
+			defer wg.Done()
+			err := s.downloadMetaInRange(start, end, batchSize, taskId)
+
+			chanRes <- err
+		}(rangeStart, rangeEnd, taskIdx, chanRes)
+	}
+
+	wg.Wait()
+
+	for i := uint64(0); i < taskNum; i++ {
+		res := <- chanRes
+		if (res != nil) {
+			return res
+		}
+	}
+
+	return nil
+}
+
+func (s *StorageManager) downloadMetaInRange(from, to, batchSize, taskId uint64) error {
+	rangeStart := from
+	for from < to {
+		s.mu.Lock()
+		localL1 := s.localL1
+		lastKvIdx := s.lastKvIdx
+		s.mu.Unlock()
+
+		batchLimit := from + batchSize
+		if batchLimit > to {
+			batchLimit = to
+		}
+		// In case remove is supported and lastKvIndex is decreased
+		if batchLimit > lastKvIdx {
+			batchLimit = lastKvIdx
+		}
+
+		kvIndices := []uint64{}
+		for i := from; i < batchLimit; i++ {
+			kvIndices = append(kvIndices, i)
+		}
+
+		metas, err := s.l1Source.GetKvMetas(kvIndices, localL1)
+		for retryTimes := 0; (retryTimes < 10) && (err != nil); retryTimes++ {
+			// Retry the request for 10 times in case it could fail occasionally in poor network connection
+			time.Sleep(2 * time.Second)
+			metas, err = s.l1Source.GetKvMetas(kvIndices, localL1)
+		}
+
+		if err != nil {
+			return err
+		}
+
+		s.mu.Lock()
+		if localL1 != s.localL1 {
+			s.mu.Unlock()
+			continue
+		}
+		for i, meta := range(metas) {
+			s.blobMetas[kvIndices[i]] = meta
+		}
+		s.mu.Unlock()
+
+		log.Info(
+			"One batch metas has been downloaded", "first", from,
+			"batchLimit", batchLimit,
+			"to", to,
+			"progress", fmt.Sprintf("%.1f%%", float64((from - rangeStart)*100)/float64(to - rangeStart)),
+			"taskId", taskId)
+
+		from = batchLimit
+	}
+	return nil
+}
+
+// This function is only called by DownloadFinished which already uses s.mu to protect the s.blobMetas, so
+// we don't need to lock in this function
+func (s *StorageManager) updateLocalMetas(kvIndices []uint64, commits []common.Hash, oldLastKvIndex uint64) {
+	for i, idx := range kvIndices {
+		meta := [32]byte{}
+		new(big.Int).SetInt64(int64(idx)).FillBytes(meta[0:5])
+		copy(meta[32-HashSizeInContract:32], commits[i][0:HashSizeInContract])
+
+		s.blobMetas[idx] = meta
+	}
+
+	// In case the lastKvIdx is smaller than oldLastKvIdx because of removal, we need to reset the removal metas
+	for i := s.lastKvIdx; i < oldLastKvIndex; i++ {
+		meta := [32]byte{}
+		new(big.Int).SetInt64(int64(i)).FillBytes(meta[0:5])
+
+		s.blobMetas[i] = meta
+	}
+}
+
+// Please note that the caller function must uses s.mu to protect the s.blobMetas reading in this function
+func (s *StorageManager) getKvMetas(kvIndices []uint64) ([][32]byte, error) {
+	metas := [][32]byte{}
+	for _, i := range kvIndices {
+		meta, ok := s.blobMetas[i]
+		if ok {
+			metas = append(metas, meta)
+		} else {
```

Comment:
I think we should do more sanity checks here:
- If i >= lastKvIdx, then return empty metas
- else return error (which is not expected to happen).